### PR TITLE
Project 08: Integration Tests and N+1 Query Optimization

### PR DESCRIPTION
**Pull request by GhausAliKhan**
### In this branch we did:
Introduced comprehensive integration tests for the Blog App, using Capybara, and addressed the N+1 query problem in data fetching.

- Developed integration tests for all views in the Blog App, ensuring the desired user experience and workflow.
- Optimized data queries to solve the N+1 problem, particularly for fetching posts and their comments.
- Followed best practices for Ruby and Rails development as outlined in Microverse's guidelines.

### Changes Made

- Integration tests added in spec/features, covering:
  - User index, show, and post index pages.
  - Individual post show page functionalities.
- Query optimizations implemented in relevant controllers and models to efficiently fetch data with eager loading.
- Ensured code readability and maintainability throughout the application.

### Testing and Logs

- Integration tests can be run using `bundle exec rspec`.
- Manual testing conducted to complement automated tests.
- Log screenshots included to demonstrate query optimization results (Before and After).

### Screenshots
_Posts.index Before_
posts index(BEFORE)

_Posts.index After_
posts index(AFTER)

### CODE REVIEW

File: spec/features/user_show_posts_spec.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+RSpec.feature 'User Posts Page', type: :feature do
```
Comment (zeff96) [inline spec/features/user_show_posts_spec.rb:3]: - [x] Hey there! 👋 After carefully reviewing the test coverage for the user show page, I've observed that certain test cases, as highlighted in the project requirements screenshots, are currently missing. 🧐 To ensure a comprehensive coverage, please incorporate these indicated test cases into your implementation. 🚀 Your attention to this matter will greatly enhance the robustness of our testing approach. Thank you! 🙌

Screenshot from 2023-12-05 17-49-43

- [x] Hey! 👋 Just a heads up, while you've successfully included the missing test case in user_show_details_spec.rb, it's not needed for this milestone. 🚫 Our focus should be on the feature test cases for user#index, user#show, post#index, and post#show views only. 🔍 Please remove any other feature test case files beyond this scope and ensure that corresponding test cases are relocated to the appropriate files. 📂 Your cooperation in streamlining the test cases will contribute to a more organized and efficient testing suite. Thanks! 🌟

File: spec/features/user_posts_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+require 'rails_helper'
+RSpec.feature 'User Post Index Page', type: :feature do
```
Comment (zeff96) [inline spec/features/user_posts_index_spec.rb:3]: - [x] Hi @Team! 👋 Fantastic job on incorporating feature tests for the user post index page! 🎉 However, I've noticed that some test cases, as indicated in the screenshot below, are either missing or incorrectly placed in the wrong file. 🤔 Kindly ensure you add the missing ones or relocate the ones placed erroneously from other files to this specific file. 🔄 Your attention to this detail will help maintain a comprehensive and organized testing suite. Much appreciated! 🙌

Screenshot from 2023-12-05 18-20-46

**Review by zeff96 — CHANGES_REQUESTED** [review]
### Status: CHANGES REQUIRED ♻️
Hi @GhausAliKhan and @78fatemanazari , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
:heavy_check_mark: Fixed n+1 queries
:heavy_check_mark: Fair workload distribution

## Required Changes ♻️

- [x] Hello @GhausAliKhan and @78fatemanazari! 👋 Just a quick note: in this milestone, your task is to write feature tests exclusively for the following files – user#index, user#show, post#index, and post#show. 📄🚀 Any files outside of this list should be deleted. 🗑️ If those files happen to include test cases that can be relocated to the aforementioned files, please ensure a smooth transfer. 🔄 Your cooperation in keeping the scope focused will contribute to an efficient and organized testing process. Thanks a bunch! 🌟

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by 1GORDON — APPROVED** [review]
# Status: APPROVED 🟢 

Hi @GhausAliKhan, and @78fatemanazari ✋

Your project is complete!  There is nothing else to say other than... it's time to merge it: ship it:
Congratulations! :tada: :clap: :clap: :100:

## Highlights!
- All tests are passing ✔️
- Added missing tests ✔️

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend considering them as they can improve your code._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@1GORDON` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**